src/shogun/lib/type_case.h

```diff
@@ -0,0 +1,170 @@
+/*
+ * This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Authors: Gil Hoben
+ */
+
+#ifndef SHOGUN_TYPE_CASE_H
+#define SHOGUN_TYPE_CASE_H
+
+#include <typeindex>
+#include <unordered_map>
+
+#include <shogun/lib/any.h>
+#include <shogun/lib/type_list.h>
+
+using namespace shogun;
+
+namespace shogun
+{
+	typedef Types<
+	    bool, char, int8_t, uint8_t, int16_t, uint16_t, int32_t, uint32_t,
+	    int64_t, uint64_t, float32_t, float64_t, floatmax_t>::type SG_TYPES;
```

Comment:
so this collects all types that we support going from any to compile time type right?
So if we added SGMatrix<float64_t>, it would be added here or?

Reply:
Yes, it should be added there. And then in the enum class TYPE and use `SG_PRIMITIVE_TYPE` to map the type to the enum and then to the appropriate map!

Reply:
can we have one vector/matrix type for illustration?

Reply:
On the other hand that might be ok as in shogun, we usually only template API function according to the primitive type, not the container type (apart from some Any things)

Reply:
Might have to think for it for a while... I cant think of a way how the compiler could do such a check at compile time. Might have to stick to the more naive implementation for now

Reply:
No I think it is good to offer functions to deal with vector/matrix/std::vector separately... we do this specialization inside Any as well (see equals/clone)

Reply:
Ah ok! I do think I found a solution. Can basically have different parameters in the sg_for_each_type to execute lambdas for specific datatypes, and then with some SFINAE magic can execute the right lambda or throw an error (or do nothing). Could a have a function signature like this:

```
template <typename Lambda1 = std::nullptr_t, typename Lambda2 = std::nullptr_t, typename Lambda3 = std::nullptr_t>
void sg_for_each_type(const Any& any, const typemap& typesmap,
            Lambda1 scalar_func = nullptr, Lambda2 vector_func = nullptr, Lambda3 matrix_func = nullptr)
```

Reply:
++ for static_assert, always nicer